File: 02. Algorithms/01. Arrays/21. Rotate the 2D Array/Rotate the 2D array.cpp
#include <iostream>

using namespace std;

int main()
{
    int a[3][3]={1,2,3,4,5,6,7,8,9};
    int i,j;
    cout<<("Original Array\n");
    for(i=0;i<3;i++)
    {
        for(j=0;j<3;j++)
        {
            cout<<("%d ",a[i][j]);
        }
        cout<<("\n");
    }
    cout<<("Rotate Matrix by 90 degrees\n");
    for(i=0;i<3;i++)
    {
        for(j=2;j>=0;j--)
        {
            cout<<("%d ",a[j][i]);
        }
        cout<<("\n");
    }
    return 0;
}

File: 02. Algorithms/01. Arrays/21. Rotate the 2D Array/Rotate the 2D array.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function(matrix) {
     let left = 0; 
    let right = matrix.length  - 1;

    while(left < right) {
        

        for(let i = 0; i < (right - left); i++) {
            // save the top left first
          let top = left;
          let bottom = right;
          const topLeft = matrix[top][left + i];
        
          matrix[top][left + i] = matrix[bottom - i][left];
          matrix[bottom - i][left] = matrix[bottom][right - i];
          matrix[bottom][right - i] = matrix[top + i][right]; 
          matrix[top + i][right] = topLeft;
        }

        right--;
        left++;
            
    }

    return matrix;
};

